modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -0,0 +1,582 @@
+// Third party copyrights are property of their respective owners.
+
+#include "../precomp.hpp"
+
+#ifdef HAVE_PROTOBUF
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <limits>
+#include <algorithm>
+
+
+#if defined(__GNUC__) && __GNUC__ >= 5
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsuggest-override"
+#endif
+#include "opencv-onnx.pb.h"
+#if defined(__GNUC__) && __GNUC__ >= 5
+#pragma GCC diagnostic pop
+#endif
+
+namespace cv {
+namespace dnn {
+CV__DNN_EXPERIMENTAL_NS_BEGIN
+
+
+class ONNXImporter
+{
+    opencv_onnx::ModelProto model_proto;
+
+    std::map<std::string, Mat> getGraphTensors(
+                                    const opencv_onnx::GraphProto& graph_proto);
+    Mat getBlob(const opencv_onnx::NodeProto& node_proto, const std::map<std::string, Mat>& constBlobs, int index);
+
+    LayerParams getLayerParams(const opencv_onnx::NodeProto& node_proto);
+
+public:
+
+    ONNXImporter(const char *onnxFile)
+    {
+        std::fstream input(onnxFile, std::ios::in | std::ios::binary);
+
+        if (!model_proto.ParseFromIstream(&input))
+            CV_Error(Error::StsUnsupportedFormat, "Failed to parse onnx model");
+    }
+
+    void populateNet(Net dstNet);
+};
+
+inline void replaceLayerParam(LayerParams& layerParams, const String& oldKey, const String& newKey)
+{
+    if (layerParams.has(oldKey)) {
+        layerParams.set(newKey, layerParams.get(oldKey));
+        layerParams.erase(oldKey);
+    }
+}
+
+void releaseONNXTensor(opencv_onnx::TensorProto& tensor_proto)
+{
+    if (!tensor_proto.raw_data().empty()) {
+        delete tensor_proto.release_raw_data();
+    }
+}
+
+Mat getMatFromTensor(opencv_onnx::TensorProto& tensor_proto)
+{
+    CV_Assert(!tensor_proto.raw_data().empty() || !tensor_proto.float_data().empty()
+                    || !tensor_proto.double_data().empty() || !tensor_proto.int64_data().empty());
+
+    opencv_onnx::TensorProto_DataType datatype = tensor_proto.data_type();
+    Mat blob;
+    std::vector<int> sizes;
+    for (int i = 0; i < tensor_proto.dims_size(); i++) {
+            sizes.push_back(tensor_proto.dims(i));
+    }
+    if (datatype == opencv_onnx::TensorProto_DataType_FLOAT) {
+
+        if (!tensor_proto.float_data().empty()) {
+            const ::google::protobuf::RepeatedField<float>& field = tensor_proto.float_data();
+            CV_Assert(!field.empty());
+            Mat(sizes, CV_32FC1, (void*)field.data()).copyTo(blob);
+        }
+        else {
+            char* val = const_cast<char*>(tensor_proto.raw_data().c_str());
+            Mat(sizes, CV_32FC1, val).copyTo(blob);
+        }
+    }
+    else if (datatype == opencv_onnx::TensorProto_DataType_DOUBLE)
+    {
+        const ::google::protobuf::RepeatedField<double>& field = tensor_proto.double_data();
+        CV_Assert(!field.empty());
+        Mat(sizes, CV_64FC1, (void*)field.data()).copyTo(blob);
+        blob.convertTo(blob, CV_32FC1);
```

Comment:
It's better to call only `convertTo` here without in-place conversions:
```cpp
Mat(sizes, CV_64FC1, (void*)field.data()).convertTo(blob, CV_32FC1);
```

Reply:
done